tools/flang1/flang1exe/dtypeutl.c

```diff
@@ -2115,6 +2115,50 @@ same_dtype(DTYPE d1, DTYPE d2)
   return FALSE;
 }
 
+/** \brief Similar to eq_dtype(), except all scalar integer types are compatible
+    with each other, all scalar logical types are compatible with each other,
+    all character types are compatible with each other and we allow
+    casting arrays when types fit into each other.
+ */
+LOGICAL
+cmpat_dtype_array_cast(DTYPE d1, DTYPE d2)
+{
+  int s1, s2;
+
+  if (d1 == d2)
+    return TRUE;
+  if (DTY(d1) != DTY(d2)) {
+    /* check for any logical first since logical types also have the
+     * _TY_INT attribute.
+     */
+    if (DT_ISLOG(d1) && DT_ISLOG(d2))
+      return TRUE;
+    if (DT_ISINT(d1) && DT_ISINT(d2))
+      return TRUE;
+    return FALSE;
+  }
+  if (DTY(d1) == TY_CHAR || DTY(d1) == TY_NCHAR)
+    return TRUE;
```

Comment:
Does this mean that any type can be successfully cast to `CHARACTER` type (with truncation)?

Reply:
To me this means that "all character types are compatible with each other". In the lines above we have already established that the types are different and that at least one of them is not logical and not an integer.
I took this line and the lines above from `cmpat_dtype()` [function body](https://github.com/flang-compiler/flang/blob/442d797075f4b5cd461593a2e216d0f49cb1eba5/tools/flang1/flang1exe/dtypeutl.c#L1889) and built on top of that similarly to what `cmpat_dtype_with_size()` has done.
Also see [the comment](https://github.com/flang-compiler/flang/blob/442d797075f4b5cd461593a2e216d0f49cb1eba5/tools/flang1/flang1exe/dtypeutl.c#L1930) which `cmpat_dtype_with_size()` has added before this check.

Reply:
Thanks for the detailed explanation; I was not reading carefully enough.